File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-16.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        int n{static_cast<int>(preorder.size())};
        unordered_map<int, int> poz;
        for (int i{}; i < n; ++i) {
            poz[inorder[i]] = i;
        }
        int k{};
        auto dfs = [&](auto self, int l, int r) -> TreeNode*{
            if (l > r) return nullptr;
            auto head{new TreeNode{preorder[k ++]}};
            head->left = self(self, l, poz[head->val] - 1);
            head->right = self(self, poz[head->val] + 1, r);
            return head;
        };
        return dfs(dfs, 0, n - 1);
    }
};

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-16.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int kthSmallest(TreeNode root, int k) {
        TreeNode node = root;
        while (node != null) {
            if (node.left == null) {
                k --;
                if (k == 0) return node.val;
                node = node.right;
            }
            else {
                TreeNode pred = node.left;
                while (pred.right != null && pred.right != node) {
                    pred = pred.right;
                }
                if (pred.right == null) {
                    pred.right = node;
                    node = node.left;
                }
                else {
                    pred.right = null;
                    k--;
                    if (k == 0) return node.val;
                    node = node.right;
                }
            }
        }
        return -1;
    }
}
